src/stirling/source_connectors/tcp_stats/tcp_stats_connector.cc

```diff
@@ -0,0 +1,113 @@
+
+using ProbeType = bpf_tools::BPFProbeAttachType;
+const auto kProbeSpecs = MakeArray<bpf_tools::KProbeSpec>(
+    {{"tcp_sendmsg", ProbeType::kEntry, "probe_entry_tcp_sendmsg", /*is_syscall*/ false},
+     {"tcp_sendmsg", ProbeType::kReturn, "probe_ret_tcp_sendmsg", /*is_syscall*/ false},
+     {"tcp_cleanup_rbuf", ProbeType::kEntry, "probe_entry_tcp_cleanup_rbuf", /*is_syscall*/ false},
+     {"tcp_retransmit_skb", ProbeType::kEntry, "probe_entry_tcp_retransmit_skb",
+      /*is_syscall*/ false}});
+
+void HandleTcpEvent(void* cb_cookie, void* data, int /*data_size*/) {
+  auto* connector = reinterpret_cast<TCPStatsConnector*>(cb_cookie);
+  auto* event = reinterpret_cast<struct tcp_event_t*>(data);
+  connector->AcceptTcpEvent(*event);
+}
+
+void TCPStatsConnector::AcceptTcpEvent(const struct tcp_event_t& event) {
+  events_.push_back(event);
+}
+
+void HandleTcpEventLoss(void* /*cb_cookie*/, uint64_t /*lost*/) {
+  // TODO(RagalahariP): Add stats counter.
+}
+
+const auto kPerfBufferSpecs = MakeArray<bpf_tools::PerfBufferSpec>({
+    {"tcp_events", HandleTcpEvent, HandleTcpEventLoss, kPerfBufferPerCPUSizeBytes,
+     bpf_tools::PerfBufferSizeCategory::kData},
+});
+
+Status TCPStatsConnector::InitImpl() {
+  sampling_freq_mgr_.set_period(kSamplingPeriod);
+  push_freq_mgr_.set_period(kPushPeriod);
+  PX_RETURN_IF_ERROR(InitBPFProgram(tcpstats_bcc_script));
+  PX_RETURN_IF_ERROR(AttachKProbes(kProbeSpecs));
+  PX_RETURN_IF_ERROR(OpenPerfBuffers(kPerfBufferSpecs, this));
+  LOG(INFO) << absl::Substitute("Number of kprobes deployed = $0", kProbeSpecs.size());
+  LOG(INFO) << "Probes successfully deployed.";
+  return Status::OK();
+}
+
+Status TCPStatsConnector::StopImpl() {
+  Close();
+  return Status::OK();
+}
+
+void TCPStatsConnector::TransferDataImpl(ConnectorContext* /* ctx */) {
+  if (FLAGS_JsonOutput != true) {
+    /* TODO: Support other output formats. */
+    return;
+  }
+  DCHECK_EQ(data_tables_.size(), 3U) << "Only three tables are allowed per TCPStatsConnector.";
+
+  PollPerfBuffers();
+
+  rapidjson::Document document;
+  document.SetObject();
+  rapidjson::Document::AllocatorType& allocator = document.GetAllocator();
+  rapidjson::Value metricsArray(rapidjson::kArrayType);
+
+  json_output::CreatePerfRecords(allocator, metricsArray, events_);
+
+  rapidjson::Value data(rapidjson::kObjectType);
+  data.AddMember(json_output::StringRef(json_output::metrics_str), metricsArray.Move(), allocator);
+  document.AddMember(json_output::StringRef(json_output::version_str),
+                     json_output::StringRef(json_output::version_value), allocator);
+  rapidjson::Value dataArray(rapidjson::kArrayType);
+  dataArray.PushBack(data.Move(), allocator);
+  document.AddMember(json_output::StringRef(json_output::data_str), dataArray.Move(), allocator);
+
+  rapidjson::StringBuffer strbuf;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(strbuf);
+  document.Accept(writer);
+  std::cout << strbuf.GetString() << std::endl;
+
```

Comment:
Here we should populate the data tables and remove the json output. Let's keep this code on the side somewhere. Let's figure out how exactly we write out json (or how to connect the data some other way) in the next PR.

Reply:
Moved Json output related code to local branch .